File: Multithreading/Thread Pool/Java/CustomThreadPool.java
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

class Worker extends Thread {
    private final BlockingQueue<Runnable> queue;

    public Worker(BlockingQueue<Runnable> queue) {
        this.queue = queue;
    }

    @Override
    public void run() {
        try {
            while (!Thread.currentThread().isInterrupted()) {
                Runnable task = queue.take();
                task.run();
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

class ThreadPool {
    private final BlockingQueue<Runnable> taskQueue;
    private final Worker[] workers;

    public ThreadPool(int numTasks) {
        this.taskQueue = new LinkedBlockingQueue<>();
        this.workers = new Worker[numTasks];

        for(int i = 0; i < numTasks; i++) {
            workers[i] = new Worker(taskQueue);
            workers[i].setName("Worker - " + (i + 1));
            workers[i].start();
        }
    }

    public void submitTask(Runnable task) {
        try {
            taskQueue.put(task);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public void shutDown() {
        for(Worker worker : workers) {
            worker.interrupt();
        }
    }
}

public class CustomThreadPool {
    public static void main(String[] args) throws InterruptedException {
        ThreadPool threadPool = new ThreadPool(3);

        for(int i = 1; i <= 10; i++) {
            final int taskID = i;
            threadPool.submitTask(()-> {
                System.out.println("Executing Task " + taskID + " via " + Thread.currentThread().getName());
                try {
                    Thread.sleep(500);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
        }

        Thread.sleep(3000);

        threadPool.shutDown();
    }
}

File: Multithreading/Thread Pool/Python/custom_thread_pool.py
import queue
import threading
import time


class Worker(threading.Thread):
    def __init__(self, worker_id: int, taskQueue: queue.Queue):
        super().__init__(daemon=False)
        self.worker_id = worker_id
        self.taskQueue = taskQueue

    def run(self):
        while True:
            func, args, kwargs = self.taskQueue.get()

            if func is None:
                self.taskQueue.task_done()
                break

            func(*args, **kwargs)
            self.taskQueue.task_done()

class ThreadPool:
    def __init__(self, numTasks: int):
        self.taskQueue = queue.Queue()
        self.workers = []

        for i in range(numTasks):
            worker = Worker(worker_id=i, taskQueue=self.taskQueue)
            worker.start()
            self.workers.append(worker)

    def submit(self, func, *args, **kwargs):
        self.taskQueue.put((func, args, kwargs))

    def shut_down(self):
        self.taskQueue.join()

        for _ in self.workers:
            self.taskQueue.put((None, [], {}))

        for worker in self.workers:
            worker.join()

def task(taskID: int):
    print(f"Doing task '{taskID}'")
    time.sleep(1)
    print(f"task '{taskID}' completed")

threadPool = ThreadPool(numTasks=3)

threadPool.submit(task, "taskID1")
threadPool.submit(task, "taskID2")
threadPool.submit(task, "taskID3")
threadPool.submit(task, "taskID4")
threadPool.submit(task, "taskID5")
threadPool.submit(task, "taskID6")
threadPool.submit(task, "taskID7")

threadPool.shut_down()